### DISCUSSION TITLE
STL smart pointers and TCL

### DISCUSSION THREAD

**Discussion by hdnn** [swig/swig#2152] [Q&A] [answered]
Hi! I am trying to wrap std::shared_ptr and std::unique_ptr to TCL without souccess.
From documentation I know next: 
1) **9.4 STL/C++ Library**: "_table with supported STL classes_" ... _some language modules support a subset of the above and some support additional STL classes. Please look for the library files in the appropriate language library directory._ - In TCL folder i don't see the std_shared_ptr.i file, is it mean, that it is not supported for this language?
2) **6.24 Smart pointers and operator->()** :... _Typically a smart-pointer is defined by a template class where the -> operator has been overloaded._ - It's true for STL smart pointers, but when I try to write something like this:
```
#ifdef SWIG
namespace std {
 template <class Type>
 class shared_ptr { //got this declaration from general shared_ptr.i file
 };
}
%template(std_i_test) std::shared_ptr<i_test>;
#endif

class i_test {
public:
 virtual std::string get_string() const = 0;
 virtual char* get_rstring() = 0;
 int x;
};
typedef std::shared_ptr<i_test> std_i_test;

std_i_test smart_ptr_test(); 
```
SWIG generate std_i_test object with empty methods list, so I can't call **get_rstring** method from the object returned by **smart_ptr_test()** function in TCL.

What am I doing wrong?

**Comment by wsfulton** [comment, marked as answer]
Hi. Correct, there is no shared_ptr support yet in TCL. A TCL user will need to contribute it. The approach to take is to copy Lib/python/boost_shared_ptr.i and Lib/python/std_shared_ptr.i into Lib/tcl/boost_shared_ptr.i and modify the typemaps for TCL. Copy the Examples/test-suite/python/li_boost_shared_ptr_runme.py to Examples/test-suite/tcl/li_boost_shared_ptr_runme.tcl and convert this runtime test to TCL. Create a Github pull request to add to SWIG.

For point 2, please read the [Smart pointers and operator->()](http://swig.org/Doc4.0/SWIGPlus.html#SWIGPlus_smart_pointers) chapter in the documentation. You're missing the `operator->`.
